### PR TITLE
Rle for greyscale lcd

### PR DESCRIPTION
**Pull request by gagarinlg**
Summary of changes:
implements RLE for images on 212x64 pixel radios to reduce binary size to make  space for serial port refactoring.

### CODE REVIEW

File: companion/src/storage/firmwareinterface.cpp [hunk 1]
```diff
@@ -220,7 +344,7 @@ void FirmwareInterface::seekSplash()
     return;
   }
 
-  if (seekSplash(QByteArray(OTX_SPS_TARANIS, OTX_SPS_SIZE), QByteArray(OTX_SPE, OTX_SPE_SIZE), 6784)) {
+  if (seekSplash(QByteArray(OTX_SPS_TARANIS, OTX_SPS_SIZE), QByteArray(OTX_SPE, OTX_SPE_SIZE), 3070)) {
```
Comment (raphaelcoeffic) [inline companion/src/storage/firmwareinterface.cpp:347]: Can we please have a `define` for that magic number?

File: companion/src/storage/firmwareinterface.cpp [hunk 2]
```diff
@@ -256,6 +380,12 @@ bool FirmwareInterface::setSplash(const QImage & newsplash)
         b[idx] |= z;
       }
     }
+    if(splashWidth == 212 && splashHeight == 64)
+    {
+	    splashSize = RleBitmap::encode(b, SPLASH_SIZE_MAX);
+	    if(splashSize > 3072)
```
Comment (raphaelcoeffic) [inline companion/src/storage/firmwareinterface.cpp]: Magic numbers 😬

Comment (gagarinlg) [reply companion/src/storage/firmwareinterface.cpp]: Yes and I do not like them.
Those, the max splash size define and the BltBitmap class are the reason for my question for a shared place for code and definitions between radio and companion

File: companion/src/storage/firmwareinterface.cpp (hunk 1 above)
Comment (gagarinlg) [reply companion/src/storage/firmwareinterface.cpp:347]: There is one, but that is not shared between companion and radio.
Just tell me where to put it and I will update the PR

File: companion/src/storage/firmwareinterface.cpp (hunk 2 above)
Comment (gagarinlg) [reply companion/src/storage/firmwareinterface.cpp]: found some defines and created a new one, but they are still not shared between radio and companion

**Comment by pfeerick** [comment]
I'm running the Companion build for this PR, and it looks like the edit screenshot function of Companion tried to load the image into the firmware, and mangled it...

**Comment by pfeerick** [comment]
Top middle (Taranis) PNG from this page => https://www.open-tx.org/screens-taranis.html when loaded into matching firmware and Companion from this PR became this abstract art: 
image

**Comment by gagarinlg** [comment]
That is unfortunate. Read/write/read back with companion worked for me and the simulator also worked. I will take a look.

**Comment by pfeerick** [comment]
Read/Write/read back is also fine for me... it can read back the image that was written to the firmware without issue... it's the actual hardware it's failing on... 😢 

These are two of the firmwares edited by companion... 00018 is currently showing nothing at all, and 00002 is the garbled one. I included 00017.png as Companion refused to save that one. 

Whereas neither OTX Companion and ETX Companion 2.5.0 had any issue writing that image to respective firmware (or showing properly).
[X9D+2019.zip](https://github.com/EdgeTX/edgetx/files/7600562/X9D%2B2019.zip)

**Comment by pfeerick** [comment]
Given the fact we've just required these radios to have SD card because of yaml, I wonder if we can actually do away with this entirely, and load splash screen from SD, just like with the colourlcd?

**Comment by raphaelcoeffic** [comment]
> Given the fact we've just required these radios to have SD card because of yaml, I wonder if we can actually do away with this entirely, and load splash screen from SD, just like with the colourlcd?

I’m afraid there is an issue with the RAM, as this happens while loading tons of other things, so that we would need to load directly from SD card I to the screen buffer. It’s worth trying, but it might be very slow.

**Comment by gagarinlg** [comment]
> Given the fact we've just required these radios to have SD card because of yaml, I wonder if we can actually do away with this entirely, and load splash screen from SD, just like with the colourlcd?

I had the same idea, when Ineead your message. 

> I’m afraid there is an issue with the RAM, as this happens while loading tons of other things, so that we would need to load directly from SD card I to the screen buffer. It’s worth trying, but it might be very slow.

Since it is just a few kB

File: radio/src/gui/212x64/splash.cpp [hunk 3]
```diff
@@ -31,7 +31,7 @@ struct {
   const uint8_t __magic_suffix[4] = { 'S','P','E',0 };
 } splash_struct;
 static_assert(sizeof(splash_struct.splashdata) <= MAXIMUM_SPLASH_IMAGE_SIZE, "");
-const unsigned char * const splash_lbm = splash_struct.splashdata+4;
```
Comment (raphaelcoeffic) [inline radio/src/gui/212x64/splash.cpp]: The first 4 bytes do not represent the dimensions anymore?

**Comment by pfeerick** [comment]
Much better now. Double error message is preferable to no clue whatsoever

https://user-images.githubusercontent.com/5500713/144143680-f23656e4-725b-4411-9f9c-c85076ebba74.mp4

.

File: radio/src/gui/212x64/splash.cpp (hunk 3 above)
Comment (gagarinlg) [reply radio/src/gui/212x64/splash.cpp]: It is only two bytes for the dimensions and that is handled in the RleBitmap class. The +4 was for the magic, but with the strict it is not needed.

**Comment by gagarinlg** [comment]
> Much better now. Double error message is preferable to no clue whatsoever what just went wrong. And the second message reinforces nothing really happened ;)

I tried to have just one message, but then I would have to throw an exception, which ist not done in that code yet, or return an enum and check everywhere.

**Comment by pfeerick** [comment]
Yeah, for now it was better to simply have *something*, without having to re-write whole portions of the code just to handle better error messaging. That can be part of 2.7+ when we're *bored*. At least it should only happen with some of the bigger images

**Pull request merged**